xds/internal/xdsclient/authority.go

```diff
@@ -32,10 +58,196 @@ import (
+	watchExpiryTimeout time.Duration                 // Resource watch expiry timeout
+	logger             *grpclog.PrefixLogger
+
+	// A two level map containing the state of all the resources being watched.
+	//
+	// The first level map key is the ResourceType (Listener, Route etc). This
+	// allows us to have a single map for all resources instead of having per
+	// resource-type maps.
+	//
+	// The second level map key is the resource name, with the value being the
+	// actual state of the resource.
+	resourcesMu sync.Mutex
+	resources   map[xdsresource.Type]map[string]*resourceState
+}
+
+func (a *authority) handleResourceUpdate(resourceUpdate transport.ResourceUpdate) error {
+	rType := a.resourceTypeGetter(resourceUpdate.URL)
+	if rType == nil {
+		return xdsresource.NewErrorf(xdsresource.ErrorTypeResourceTypeUnsupported, "Resource URL %v unknown in response from server", resourceUpdate.URL)
+	}
+
+	opts := &xdsresource.DecodeOptions{
+		BootstrapConfig: a.bootstrapCfg,
+		Logger:          a.logger,
+	}
+	updates, md, err := decodeAllResources(opts, rType, resourceUpdate)
+	a.updateResourceStateAndScheduleCallbacks(rType, updates, md)
+	return err
+}
+
+func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Type, updates map[string]resourceDataErrTuple, md xdsresource.UpdateMetadata) {
+	a.resourcesMu.Lock()
+	defer a.resourcesMu.Unlock()
+
+	resourceStates := a.resources[rType]
+	for name, uErr := range updates {
+		if state, ok := resourceStates[name]; ok {
+			// Cancel the expiry timer associated with the resource once a
+			// response is received, irrespective of whether the update is a
+			// good one or not.
+			state.wTimer.Stop()
+			state.wState = watchStateRespReceived
+
+			if uErr.err != nil {
+				// On error, keep previous version of the resource. But update
+				// status and error.
+				state.md.ErrState = md.ErrState
+				state.md.Status = md.Status
+				for watcher := range state.watchers {
+					watcher := watcher
+					err := uErr.err
+					a.serializer.Schedule(func(context.Context) { watcher.OnError(err) })
+				}
+				continue
+			}
+			// If we get here, it means that the update is a valid one. Notify
+			// watchers only if this is a first time update or it is different
+			// from the one currently cached.
+			if state.cache == nil || !state.cache.Equal(uErr.resource) {
+				for watcher := range state.watchers {
+					watcher := watcher
+					resource := uErr.resource
+					a.serializer.Schedule(func(context.Context) { watcher.OnUpdate(resource) })
+				}
+			}
+			// Sync cache.
+			a.logger.Debugf("Resource type %q with name %q, value %s added to cache", rType.TypeEnum().String(), name, uErr.resource.ToJSON())
+			state.cache = uErr.resource
+			// Set status to ACK, and clear error state. The metadata might be a
+			// NACK metadata because some other resources in the same response
+			// are invalid.
```

Comment:
It's not clear to me why the fact that other resources were invalid affects the state for this resource (specifically, why is Version set specially?)

Reply:
Why are there two places where (the same?) `Version` is stored in an update?  That seems to confuse things and makes some of the processing more complicated.

Reply:
SGTM.  I think that probably makes more sense, too, and seems fine to save it for a future PR.